lib/evmone/eof.cpp

```diff
@@ -90,37 +166,285 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+
+        if (t.outputs_num > OUTPUTS_INPUTS_NUMBER_LIMIT ||
+            t.inputs_num > OUTPUTS_INPUTS_NUMBER_LIMIT)
+            return EOFValidationError::inputs_outputs_num_above_limit;
+    }
+
+    return types;
+}
+
 EOFValidationError validate_instructions(evmc_revision rev, bytes_view code) noexcept
 {
     assert(!code.empty());  // guaranteed by EOF headers validation
 
+    const auto& cost_table = baseline::get_baseline_cost_table(rev, 1);
+
     size_t i = 0;
-    uint8_t op = code[0];
+    uint8_t op = 0;
     while (i < code.size())
     {
         op = code[i];
-        const auto& since = instr::traits[op].since;
-        if (!since.has_value() || *since > rev)
+        if (cost_table[op] == instr::undefined)
             return EOFValidationError::undefined_instruction;
 
-        i += instr::traits[op].immediate_size;
+        if (op == OP_RJUMPV)
+        {
+            if (i + 1 < code.size())
+            {
+                const auto count = code[i + 1];
+                if (count < 1)
+                    return EOFValidationError::invalid_rjumpv_count;
+                i += static_cast<size_t>(1 /* count */ + count * 2 /* tbl */);
+            }
+            else
+                return EOFValidationError::truncated_instruction;
+        }
+        else
+            i += instr::traits[op].immediate_size;
+
+        if (i >= code.size())
+            return EOFValidationError::truncated_instruction;
+
         ++i;
     }
 
-    if (!instr::traits[op].is_terminating)
-        return EOFValidationError::missing_terminating_instruction;
-
     return EOFValidationError::success;
 }
 
+bool validate_rjump_destinations(
+    const EOF1Header& header, size_t code_idx, bytes_view::const_iterator container) noexcept
+{
+    // Collect relative jump destinations and immediate locations
+    std::vector<size_t> rjumpdests;
+    const auto code_size = header.code_sizes[code_idx];
+    std::vector<bool> immediate_map(code_size);
+    for (size_t i = 0; i < code_size; ++i)
+    {
+        const auto op_pos = header.code_begin(code_idx) + i;
+        const auto op = container[op_pos];
+
+        if (op == OP_RJUMP || op == OP_RJUMPI)
+        {
+            const auto offset_hi = container[op_pos + 1];
+            const auto offset_lo = container[op_pos + 2];
+            const auto offset = static_cast<int16_t>((offset_hi << 8) + offset_lo);
+            const auto jumpdest = static_cast<int32_t>(i) + 3 + offset;
+            if (jumpdest < 0 || jumpdest >= code_size)
+                return false;
+            rjumpdests.push_back(static_cast<size_t>(jumpdest));
+        }
+        else if (op == OP_RJUMPV)
+        {
+            constexpr auto REL_OFFSET_SIZE = sizeof(int16_t);
+
+            const auto count = container[op_pos + 1];
+
+            const auto post_offset =
+                static_cast<int32_t>(1 + 1 /* count */ + count * REL_OFFSET_SIZE /* tbl */);
+
+            for (size_t k = 0; k < count; ++k)
```

Comment:
Also you could use offset as loop variable and avoid `* REL_OFFSET_SIZE` inside the loop.

Reply:
Done but I left `REL_OFFSET_SIZE` declaration because it's used 3 times.